### PR TITLE
feat: give a empty non-null empty sourcemap warnings

### LINKED ISSUE #1699: Give a empty `non-null` empty sourcemap warnings
### What problem does this feature solve?

.

### What does the proposed API look like?

See https://github.com/rollup/rollup/blob/28546b5821efcb72c2eb05f422d986524647a0e3/src/utils/transform.ts#L73-L74 for details.

### PR DESCRIPTION
**Pull request by shulaoda**
### Description
closes #1699

### CODE REVIEW

**Comment by shulaoda** [comment]
It seems that the hook `load` on the `JS` side has already performed `JSON.parse` on the map, whereas the hook `transform` and `render_chunk` directly pass strings to the `Rust` side for parsing. Should we unify this behavior?
https://github.com/rolldown/rolldown/blob/c5aafadfa5704cbe679bf6fb68e1810d08a642a3/packages/rolldown/src/plugin/bindingify-build-hooks.ts#L251-L263

File: crates/rolldown/tests/rolldown/warnings/sourcemap_broken/load/artifacts.snap [hunk 1]
```diff
@@ -0,0 +1,18 @@
+---
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/rolldown/warnings/sourcemap_broken/load/artifacts.snap]: rollup don't generate missing warning for load hook, please remove this testcase.

**Comment by IWANABETHATGUY** [comment]
Also I don't see any warning emitting logic, and I tried with an example using transform without `map` return, it did not emit any warning. Is that enough just mark them as missing like rollup?

**Comment by shulaoda** [comment]
> Also I don't see any warning emitting logic, and I tried with an example using transform without `map` return, it did not emit any warning. Is that enough just mark them as missing like rollup?

Which one doesn't emit any warnings, `Rollup` or `Rolldown`?

**Comment by IWANABETHATGUY** [comment]
Greak works. But it seems that emitting such warnings in real world might cause to emitting thousand times. We need some time to think through if this is good for users and if there are other ways to implement this PR.

**Comment by shulaoda** [comment]
> Greak works. But it seems that emitting such warnings in real world might cause to emitting thousand times. We need some time to think through if this is good for users and if there are other ways to implement this PR.

It seems that implementing it directly on the `JS` side is also possible.

**Comment by IWANABETHATGUY** [comment]
> > Greak works. But it seems that emitting such warnings in real world might cause to emitting thousand times. We need some time to think through if this is good for users and if there are other ways to implement this PR.
> 
> It seems that implementing it directly on the `JS` side is also possible.

It is not about implementation details, it is the necessity to emit warning for each plugin(which don't return sourcemap), I am hesitant to accept this pull request before having research in real-world usage. Giving a warning is fine, the hard thing is how, to emit warning for each  source is obviously not acceptable for user.

**Comment by underfin** [comment]
The hook warnings need to implement first, we could discuss it at https://github.com/rolldown/rolldown/issues/1903.

> emit warning for each source is obviously not acceptable for user

I think we could provide a API  like `waring_once` to avoid it.

**Comment by IWANABETHATGUY** [comment]
@shulaoda Marking the pull request to draft does not mean we are not willing to accept anymore, we will merge the pull request if some prerequisites are ready, and we will re-review the pr when we have some manpower, feel free to reopen it. Anyway, thanks for your contribution.

**Comment by shulaoda** [comment]
> @shulaoda Marking the pull request to draft does not mean we are not willing to accept anymore, we will merge the pull request if some prerequisites are ready, and we will re-review the pr when we have some manpower, feel free to reopen it. Anyway, thanks for your contribution.

Apologies, due to PR #2131, I conducted numerous PR tests in my forked repository, which caused my GitHub contribution record to become quite cluttered. Consequently, I deleted the forked repository to clean things up but forgot to consider that there were some draft PRs still existing. 🧎

**Comment by IWANABETHATGUY** [comment]
> > @shulaoda Marking the pull request to draft does not mean we are not willing to accept anymore, we will merge the pull request if some prerequisites are ready, and we will re-review the pr when we have some manpower, feel free to reopen it. Anyway, thanks for your contribution.
> 
> Apologies, due to PR #2131, I conducted numerous PR tests in my forked repository, which caused my GitHub contribution record to become quite cluttered. Consequently, I deleted the forked repository to clean things up but forgot to consider that there were some draft PRs still existing. 🧎

hhh, that's alright.

**Pull request closed without merging**